### DISCUSSION TITLE
Running Firedrake

### DISCUSSION THREAD

**Discussion by Toomyalla** [firedrakeproject/firedrake#2712] [Firedrake support] [answered]
Hello, 

I would like to know how codes were run in this tutorial.

https://www.archer2.ac.uk/training/courses/210823-firedrake/

I run Firedrake using Docker container on Windows, but I can not show some results easily. 

Best Regards

**Comment by connorjward** [comment, marked as answer]
Can you please provide more information about your issue? What do you mean by "can not show some results easily"?

I just tried to view Jupyter notebooks run from inside Docker and found that I needed to run the command

```bash
$ jupyter notebook --port=8888 --ip=0.0.0.0 --no-browser .
```

after setting up port forwarding when starting the container:

```bash
$ docker run -it -p 8888:8888 firedrakeproject/firedrake:latest
```

Note that you also need to activate the virtual environment and `pip install jupyter`.

**Comment by Toomyalla** [comment]
Thank you for your response

For example, I am trying to plot the sparsity of the matrix, but it does not appear in the way that I run the code. 

from firedrake import *
mesh = UnitSquareMesh(10, 10)

V = FunctionSpace(mesh, "CG", 1)

u = TrialFunction(V)
v = TestFunction(V)

f = Function(V)
x, y = SpatialCoordinate(mesh)
f.interpolate((1+8*pi*pi)*cos(x*pi*2)*cos(y*pi*2))

a = (inner(grad(u), grad(v)) + inner(u, v)) * dx
L = inner(f, v) * dx

u = Function(V)

solve(a == L, u, solver_parameters={'ksp_type': 'cg', 'pc_type': 'none'})

from firedrake.petsc import PETSc
petsc_mat = assemble(a).M.handle
import scipy.sparse as sp

indptr, indices, data = petsc_mat.getValuesCSR()
scipy_mat = sp.csr_matrix((data, indices, indptr), shape=petsc_mat.getSize())
import matplotlib.pyplot as plt
plt.spy(scipy_mat)

**Reply by connorjward** [reply]
There are a few syntax errors in this code, but after fixing those, import `pyplot as plt` and adding `plt.show()` at the end I can see a sparsity pattern.

I can imagine that `plt.show()` might not work from inside a Docker container without some tricky configuration. You may need to use `plt.savefig(...)` instead.

**Reply by connorjward** [reply]
Ah but if you're running inside Jupyter this shouldn't be a problem. I'm struggling to determine what it is you aren't able to do at the moment. Is it that the code runs but the figure isn't displayed in the notebook?

**Comment by Toomyalla** [comment]
Yes, this is the problem.

**Reply by connorjward** [reply]
I believe that the trick is to use `%matplotlib inline` ([e.g.](https://www.statology.org/matplotlib-inline/)). You could likely simplify things by just trying to display a very simple figure without using Firedrake.

**Comment by Toomyalla** [comment]
Thank you
